Ble tracker

open-homeautomation opened the pull request:
**Description:** Tracking of BLE devices. This code is similar to the Bluetooth tracker, but the code is quite different as it uses gattlib.
Gattlib is only compatible with Linux as it uses gatttool from bluez
BLE tracking can be used with iBeacons, but also other devices as fitness trackers

**Related issue (if applicable):** 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** Not yet done

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
device_tracker:
  platform: bluetooth_le_tracker
  track_new_devices: yes
  interval_seconds: 60
  consider_home: 300
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices, web services, or a:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
- [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

Commit:
Added Bluetooth Low Energy device tracker

Commit:
Added new file(s)

Commit:
Fixed pylint errors

Commit:
Remove traling zeros from device names

Commit:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into ble-tracker

Commit:
recreated deleted file

Commit:
Added requirements

open-homeautomation commented:
Ok, I have a problem here. gattlib doesn't install on the development system. It works on my Linux on the Raspberry Pi, but also not on MacOS. 
What is the recommended  way to deal with a Python library that doesn't compile/install on all platforms?

balloob commented:
You can add it to this list: https://github.com/home-assistant/home-assistant/blob/dev/script/gen_requirements_all.py#L9

balloob commented:
I think you will also have to add a `# pylint: disable-import-error`

balloob commented on homeassistant/components/device_tracker/ble_tracker.py:
```diff
@@ -0,0 +1,90 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['gattlib']
```
Please specify a version.

balloob commented on homeassistant/components/device_tracker/ble_tracker.py:
```diff
@@ -0,0 +1,90 @@
+BLE_PREFIX = 'BLE_'
+
+def setup_scanner(hass, config, see):
+    """Setup the BLE Scanner."""
```
I do not like this name. It is an abbreviation but it means the same as the other tracker, that will be confusing.

We can rename it bluetooth_tracker_bluez, that way it is clear what is being used.

open-homeautomation replied:
The Bluetooth tracker and the Bluetooth Low Energy tracker are different. The Bluetooth tracker can only detect Bluetooth devices, but not BLE, while the new module can only track Bluetooth Low energy devices. 
While booth seems to be the same, Bluetooth and Bluetooth Low Energy differ a lot.
That's why I don't want to call it "bluetooth_tracker_...". If you don't want abbreviations, I can rename it to "bluetooth_le_tracker" or "bluetooth_low_energy_tracker". Personally I think both names a bit long.

balloob replied:
I am still confused. The other bluetooth tracker depends on pybluez. pygattlib claims to use the gatttool from Bluez.

Let's call it bluetooth_le_tracker.py, that way it is more clear it's still bluetooth. We should also make sure to emphasize the differences in the docs.

open-homeautomation replied:
Yes, both uses bluez, but completely different functionalities from bluez.
The APIs for Bluetooth and Bluetooth LE are completely different. 

No problem for me renaming it to Bluetooth_LE. I will write a small text that explains the differences between Bluetooth and Bluetooth LE.

Commit:
Renamed to bluetooth_le tracker
Removed gattlib from tests
Minor code cleanup

balloob commented on .coveragerc:
```diff
@@ -113,6 +113,7 @@ omit =
     homeassistant/components/device_tracker/aruba.py
     homeassistant/components/device_tracker/asuswrt.py
     homeassistant/components/device_tracker/bluetooth_tracker.py
+    homeassistant/components/device_tracker/ble_tracker.py
```
Can you move it before `bluetooth_tracker`? 🔤

balloob commented:
Ok to merge when last comment addressed 🐬

Commit:
- fixed .coveragerc bug
- changed discovery algorithm, new devices will only be added if seen 5 times to make sure
  HA doesn't blow the database with devices just passing by

DavidMStraub commented:
Hi, it's awesome to have a BLE device tracker component, I got a small ibeacon the other day and would love to use it with HASS. In fact I have a question on your new component - my apologies if a comment in the PR is not the right place to ask. So on my system - a Pi 3 with HASS installed using the [all-in-one installer](https://home-assistant.io/getting-started/installation-raspberry-pi-all-in-one/) - HASS does not run as root, so while I can install gatttlib, the component cannot be initialized  as the  scanning requires root privileges. Are you aware of any way to make it work?

Landrash commented:
@DavidMStraub It's the wrong place to ask yes. Wont stop me from trying to answer.
Bluetooth LE would currently need some escalated privileges. There's some [documentation](http://unix.stackexchange.com/a/182559) of folks fixing it for the manual hcitool so I think this could be adapted to work with HA. Can't be a out of the box fix most likely.
